### ISSUE TITLE
StatefulSet pod anti-affinity update not working

### ISSUE THREAD

**Issue by bremy11** [kubernetes/kubernetes#68101]
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

/kind bug

**What happened**:

Updating a statefulset sometimes bypasses pod anti-affinity depending on the labels applied on the pod itself.

**What you expected to happen**:
**How to reproduce it (as minimally and precisely as possible)**:

On a 3 node cluster I cover all the nodes with statefulsets i1, i2, and i3 (i1.yaml, i2.yaml, i3.yaml below) that have anti-affinity with each other based on the label "bbb" and one replica each. There is also a statefulset i0 with one replica which has label "aaa" and anti-affinity based on the label "aaa" (i0-step1.yaml below). All four of these statefulsets have their pods created successfully, with the pod of i0 colocated with one of the pods of i1/i2/i3.

Then I update statefulset i0 to also have an anti-affinity based on the label "bbb". Depending on whether I also add the label "bbb" to the pod template of i0, I see different behavior:

- If I don't add the label (i0-step2-doesntwork.yaml below), the pod restarts and reaches Running state even though it should have been prevented from running anywhere due to the anti-affinity with label "bbb".

- If I *do* add the label (i0-step2-works.yaml below), the pod restarts and stays in the Pending state as expected. Describing the pod correctly indicates that it cannot find a node to be scheduled on because of anti-affinity constraints.

The first behavior seems to be a bug, because the presence of the "bbb" label on i0's pod should not affect its anti-affinity with other pods that have that label, because the anti-affinity constraint exists regardless.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

    ```
    Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:53:20Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
    Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
    ```

- Cloud provider or hardware configuration: Three Ubuntu 16.04.5 VMs on Server 2012R2 Hyper-V host
- OS (e.g. from /etc/os-release): `VERSION="16.04.5 LTS (Xenial Xerus)"`
- Kernel (e.g. `uname -a`): `Linux brremy-ub1 4.4.0-134-generic #160-Ubuntu SMP Wed Aug 15 14:58:00 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: 
- Others:

---
<details>
<summary>Repro spec files</summary>

i0-step1.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i0
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i0
  serviceName: ""
  template:
    metadata:
      labels:
        aaa: ""
        instance: i0
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: aaa
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

i1.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i1
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i1
  serviceName: ""
  template:
    metadata:
      labels:
        bbb: ""
        instance: i1
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: bbb
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

i2.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i2
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i2
  serviceName: ""
  template:
    metadata:
      labels:
        bbb: ""
        instance: i2
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: bbb
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

i3.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i3
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i3
  serviceName: ""
  template:
    metadata:
      labels:
        bbb: ""
        instance: i3
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: bbb
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

i0-step2-doesntwork.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i0
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i0
  serviceName: ""
  template:
    metadata:
      labels:
        aaa: ""
        instance: i0
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: aaa
                operator: Exists
            topologyKey: kubernetes.io/hostname
          - labelSelector:
              matchExpressions:
              - key: bbb
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

i0-step2-works.yaml

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: i0
spec:
  replicas: 1
  selector:
    matchLabels:
      instance: i0
  serviceName: ""
  template:
    metadata:
      labels:
        aaa: ""
        bbb: ""
        instance: i0
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: aaa
                operator: Exists
            topologyKey: kubernetes.io/hostname
          - labelSelector:
              matchExpressions:
              - key: bbb
                operator: Exists
            topologyKey: kubernetes.io/hostname
      containers:
      - name: foo
        image: k8s.gcr.io/pause:2.0
```

</details>

**Comment by Huang-Wei** [comment]
/assign

It looks like a bug. I will take a look.

**Comment by bsalamat** [comment]
@Huang-Wei You are right, #68173 cannot be directly cherrypicked into 1.11. We need to develop another fix for 1.11.
